### PR TITLE
Feature/19 data preprocessing

### PR DESCRIPTION
**Pull request by ultiwinter**
Specified the numerical and categorical data for preprocessing (could be subject to change).
Added encoding functionalities for categorical data.
Added scaling options for numerical data that needs scaling.
Established pipeline for preprocessing the data.
Handles errors and exceptions.
Filters out the data that couldn't be used for Google Places API and leave the useful data for the AI predictor.
Created a preprocessing demo and added "Data preprocessing" to the main.
Runs on both historical data on S3 and local data

### CODE REVIEW

File: src/preprocessing/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        scaler = Normalizer()
+        self.preprocessed_df[column] = scaler.fit_transform(
+            self.preprocessed_df[[column]]
+        )
+        return self.preprocessed_df
+
+    def remove_outliers_zscore(self, column):
+        z_scores = stats.zscore(self.preprocessed_df[[column]])
+        self.preprocessed_df[column] = self.preprocessed_df[
+            (z_scores < 3) & (z_scores > -3)
```
Comment (rbbozkurt) [inline src/preprocessing/preprocessing.py:105]: might use top-level variables here 
```suggestion
            (z_scores < THRESHOLD) & (z_scores > -THRESHOLD)
```

File: src/preprocessing/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    MultiLabelBinarizer,
+    Normalizer,
+    OneHotEncoder,
+    RobustScaler,
+    StandardScaler,
+)
+
+current_dir = os.path.dirname(__file__) if "__file__" in locals() else os.getcwd()
+parent_dir = os.path.join(current_dir, "..")
+file_path = os.path.join(current_dir, "../data/last_snapshot.csv")
```
Comment (rbbozkurt) [inline src/preprocessing/preprocessing.py:23]: might use repository class here.

File: src/preprocessing/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        return self.preprocessed_df
+
+    def save_preprocessed_data(self):
+        columns_to_save = []
+        columns_to_save.extend(self.numerical_data)
+        columns_to_save.extend(self.added_classes)
+
+        selected_df = self.preprocessed_df[columns_to_save]
+        try:
+            save_path = os.path.join(current_dir, "../data/preprocessed_data.csv")
```
Comment (rbbozkurt) [inline src/preprocessing/preprocessing.py:179]: please also consider using repository.py here

File: src/preprocessing/preprocessing.py (hunk 2 above)
Comment (rbbozkurt) [reply src/preprocessing/preprocessing.py:23]: And where is` last_snapshot.csv` is being created?

**Review by rbbozkurt — CHANGES_REQUESTED** [review]
Good work on preprocessing! But please consider injecting an instance of repository.py or you can also directly import database by using `from database import get_database`. Last but not least, you can also extend repository.py by adding preprocessing related operations.

File: src/preprocessing/preprocessing.py (hunk 1 above)
Comment (ultiwinter) [reply src/preprocessing/preprocessing.py:105]: Addressed.

File: src/preprocessing/preprocessing.py (hunk 3 above)
Comment (ultiwinter) [reply src/preprocessing/preprocessing.py]: Addressed.

**Review by rbbozkurt — APPROVED** [review]
Lots of good work done!

File: src/preprocessing/preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+sys.path.append(current_dir)
+log = get_logger()
+
+
+class Preprocessing:
+    def __init__(self, filter_null_data=True, historical_data=False):
+        data_repo = get_database()
+        data_path = data_repo.get_output_path()
+        log.info(f"Data path = {data_path}")
```
Comment (felix-zailskas) [inline src/preprocessing/preprocessing.py:35]: maybe more of a debug statement

File: src/preprocessing/preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    def normalization(self, column):
+        if column in self.preprocessed_df.columns:
+            scaler = Normalizer()
+            self.preprocessed_df[column] = scaler.fit_transform(
+                self.preprocessed_df[[column]]
+            )
+        return self.preprocessed_df
+
+    def remove_outliers_zscore(self, column):
+        THRESHOLD = 3
```
Comment (felix-zailskas) [inline src/preprocessing/preprocessing.py:134]: I assume this corresponds to a 99% interval or is it more of an intuitive value?

**Review by felix-zailskas — COMMENTED** [review]
Runs well on my side just some small comments. Good Job!

File: src/preprocessing/preprocessing.py (hunk 4 above)
Comment (ultiwinter) [reply src/preprocessing/preprocessing.py:35]: Addressed.

File: src/preprocessing/preprocessing.py (hunk 5 above)
Comment (ultiwinter) [reply src/preprocessing/preprocessing.py:134]: It is more of a conventional, a rule-of-thumb value, it worked well before.

**Pull request merged**